### PR TITLE
Add COG children picker to forms

### LINKED ISSUE #5185: Add UI COG - Part 1 (add and visualize COG in forms)
To use the new COG feature, the user will follow this simple workflow:

1. **Open the New COG Form**:
   - Navigate to `Data Entry > COG Table` to open a new COG form.

2. **COG Form Layout**:
   - The form includes:
     - A label named "Name" with an associated input box.
     - An empty subview with a "+" button.

3. **Add COG/CO**:
   - Click the "+" button.
   - A dialog opens with the following options:
     - **Add COG**: 
       - Create a new COG.
       - Use QB (Query Builder) to search for an existing COG.
     - **Add CO**:
       - Create a new CO.
       - Use QB to search for an existing CO.

4. **Create or Select Records**:
   - The user can either create a new record and save it or select one or several records from the QB search results and add them to the COG parent.

5. **Updated Parent COG Subview**:
   - A new record line is added to the Parent COG subview. This line includes:
     - COG/CO information in a collapsed subview.
     - Attributes from COJO (e.g., isPrimary, isSubstrate).
     - A trash can icon at the end of each line to remove the record from the parent COG.
     - An arrow at the front of each line to expand the form.
     - Clicking on the record link opens a new page with the record form

6. **Expand Record Form**:
   - When the arrow is clicked to uncollapse the line, it opens the corresponding record form using a conditional format based on whether the foreign key (FK) CO or FK COG is present.
   
   
Screenshot 2024-08-05 at 12 32 05 PM
Screenshot 2024-08-05 at 12 32 13 PM
Screenshot 2024-08-05 at 12 32 23 PM
Screenshot 2024-08-05 at 12 32 31 PM
Screenshot 2024-08-05 at 12 32 39 PM

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #5185

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add automated tests
- [ ] Add relevant issue to release milestone

### Testing instructions

- go to data entry and select Collection Object Group 
- define a name and select a cogType 
- [ ] verify you can create a new CO child and save the CO 
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the CO child is still there 
- [ ] verify you can create a new COG child and save the COG 
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the COG child is still there 
- [ ] verify you can add one CO child using query builder
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the CO child is still there 
- [ ] verify you can add several CO children using query builder
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the CO child is still there
- [ ] verify you can add one COG child using query builder
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the COG child is still there 
- [ ] verify you can add several COG children using query builder
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the COG child is still there
- [ ] verify you can delete any children 
- [ ] verify you can uncollapse each child 
- you should not be able to add a COGChild to a COG that is already in use in another COG, when doing so an error should be thrown saying that the COGChild is already used 

This PR affects database migrations. See migration testing instructions ([here](https://specify.github.io/testing/pull_requests.html))

View Def  : [geo.common.views.xml.zip](https://github.com/user-attachments/files/17590977/better.common.views.xml.zip)

### CODE REVIEW

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
To make this work, I had to manually add `CollectionObjectGroupType` records.

```sql
INSERT INTO ku_fish.collectionobjectgrouptype (COGTypeID, Name, Type, Version, TimestampCreated, TimestampModified,
                                               CollectionID, CreatedByAgentID, ModifiedByAgentID)
VALUES (1, 'Consolidated', 'Consolidated', 0, '2024-10-21 14:03:04.000000', '2024-10-21 14:03:04.000000', 4, 3, 3);

INSERT INTO ku_fish.collectionobjectgrouptype (COGTypeID, Name, Type, Version, TimestampCreated, TimestampModified,
                                               CollectionID, CreatedByAgentID, ModifiedByAgentID)
VALUES (2, 'Drill Core', 'Drill Core', 0, '2024-10-21 14:03:04.000000', '2024-10-21 14:03:04.000000', 4, 3, 3);

INSERT INTO ku_fish.collectionobjectgrouptype (COGTypeID, Name, Type, Version, TimestampCreated, TimestampModified,
                                               CollectionID, CreatedByAgentID, ModifiedByAgentID)
VALUES (3, 'Discrete', 'Discrete', 0, '2024-10-21 14:03:04.000000', '2024-10-21 14:03:04.000000', 4, 3, 3);
```


> ### Testing instructions
> 
> - go to data entry and select Collection Object Group 
> - define a name and select a cogType

This could **only** be done after manually running the SQL command above. This _should_ be solved in #5215, so we need to make sure that is actually resolved.

> - [ ] verify you can create a new CO child and save the CO 

Only if I create it via the `childCojos` relationship. The `cojo` relationship does not preserve the children on the form.

https://github.com/user-attachments/assets/190d0806-f1d7-4261-8604-b7d9c2de0c2c

> - [X] verify you can save the main COG form 
> - [ ] verify after refresh the CO child is still there 

See above for `cojo`

> - [X] verify you can create a new COG child and save the COG 
> - [X] verify you can save the main COG form 
> - [] verify after refresh the COG child is still there 

See above for `cojo`, only preserved for `childCojos`

> - [X] verify you can add one CO child using query builder
> - [X] verify you can save the main COG form 
> - [ ] verify after refresh the CO child is still there 
> - [ ] verify you can add several CO children using query builder

Each time I used the QB to return a list of all COs, searching only (any) Catalog Number, I would select a range of results. After selection, an item in the list would always be duplicated (recreated several times).

https://github.com/user-attachments/assets/d1dcea72-1be9-4a12-a7c8-7278b571dd7b


[Specify 7 Crash Report - 2024-10-21T19_28_33.009Z.txt](https://github.com/user-attachments/files/17466119/Specify.7.Crash.Report.-.2024-10-21T19_28_33.009Z.txt)

This is an **existing issue**, but it does affect this component, so I've reported it here:
https://github.com/specify/specify7/issues/5335

> - [X] verify you can save the main COG form 
> - [ ] verify after refresh the CO child is still there

See above for `cojo`, only preserved for `childCojos`

> - [X] verify you can add one COG child using query builder
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the COG child is still there 
> - [X] verify you can add several COG children using query builder

Same issue reported above (#5335)

> - [X] verify you can save the main COG form 
> - [ ] verify after refresh the COG child is still there

See above for `cojo`, only preserved for `childCojos`

> - [X] verify you can delete any children 

I had some trouble with this being consistently enforced, as sometimes the children I added wouldn't be saved, but I can't recreate consistently. Would like other @specify/ux-testing members to see if they can identify the issue.

> - [X] verify you can uncollapse each child 
> - [ ] verify you can add a parent cojo 

I cannot add a parent Cojo. When I try to create a COJO, I can't set it's parent, therefore I see:
```
BusinessRuleException at /api/specify/collectionobjectgroupjoin/ Both childcog and childco cannot be null.
```

[Specify 7 Crash Report - 2024-10-21T19_47_16.967Z.txt](https://github.com/user-attachments/files/17466288/Specify.7.Crash.Report.-.2024-10-21T19_47_16.967Z.txt)

**Review by sharadsw — CHANGES_REQUESTED** [review]

File: specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  readonly collection: Collection<AnySchema> | undefined;
+}): JSX.Element | null {
+  const [isOpen, handleOpen, handleClose] = useBooleanState();
+  const COJOChildrenTables = [
+    tables.CollectionObject,
+    tables.CollectionObjectGroup,
+  ];
+  const [state, setState] = React.useState<'Add' | 'Search' | undefined>(
+    undefined
+  );
```
[the remark this answers is not in the document]
Comment (maxpatiiuk) [reply specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx:35]: I believe we are using the name `state` in several components for this purpose.
The name is generic, but useful in simple components that have only one or two states.

File: specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  const [resource, setResource] = React.useState<
+    | SpecifyTable<CollectionObject>
+    | SpecifyTable<CollectionObjectGroup>
+    | undefined
+  >(undefined);
```
[the remark this answers is not in the document]
Comment (maxpatiiuk) [reply specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx:40]: 💯

File: specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+          buttons={commonText.cancel()}
+          dimensionsKey="COGChildren"
+          header={formsText.addCOGChildren()}
+          onClose={handleClose}
+        >
+          <div className="flex flex-col gap-4">
+            {COJOChildrenTables.map((table) => (
+              <div className="flex items-center gap-2" key={table.name}>
+                <TableIcon label name={table.name} />
+                {localized(table.label)}
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/FormCells/COJODialog.tsx:104]: `localized()` probably not needed here

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
In addition to my previous review, I wanted to mention that I can not always delete an item from a COG, and after a refresh it will return:

https://github.com/user-attachments/assets/b92ea2da-5bd3-419f-a2ac-aa505290feca

See the save button does not get triggered:

https://github.com/user-attachments/assets/87de0afe-db01-4604-9af3-1fdca6d8130a

**Comment by CarolineDenis** [comment]
TODO: 
- Remove cojo on cog ? And only keep parentCojo

**Review by sharadsw — APPROVED** [review]

**Review by emenslin — CHANGES_REQUESTED** [review]
### Testing instructions

- go to data entry and select Collection Object Group 
- define a name and select a cogType 
- [x] verify you can create a new CO child and save the CO 
- [x] verify you can save the main COG form 
- [x] verify after refresh the CO child is still there 
- [x] verify you can create a new COG child and save the COG 
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there 
- [x] verify you can add one CO child using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the CO child is still there 
- [x] verify you can add several CO children using query builder
- [ ] verify you can save the main COG form 
- [ ] verify after refresh the CO child is still there
- [x] verify you can add one COG child using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there 
- [x] verify you can add several COG children using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there
- [x] verify you can delete any children 
- [x] verify you can uncollapse each child 
- [x] verify you can add a parent cojo 

--- 

Adding several CO children with the query builder throws an error when you try to save the COG. Adding just one CO child does work though 

https://github.com/user-attachments/assets/d375ea7c-554f-48a4-9328-5052c0fbcefe


[Specify 7 Crash Report - 2024-10-31T16_16_00.610Z.txt](https://github.com/user-attachments/files/17590329/Specify.7.Crash.Report.-.2024-10-31T16_16_00.610Z.txt)

Everything else appears to be working though!

**Comment by CarolineDenis** [comment]
@specify/ux-testing 
When adding several children using the QB, it can happen that a row is being duplicated.
This is a known issue https://github.com/specify/specify7/issues/5335
It will not be addressed in this pr. 
Just delete the row that have been duplicated before saving your cog

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
This migration can correctly be rolled back (`0010_updateDelete_parentcojo`): ❤️ 
```
specify@494f280856e6:/opt/specify7$ ve/bin/python manage.py migrate specify 0005_collectionobjectgroup_parentcojo
Operations to perform:
  Target specific migration: 0005_collectionobjectgroup_parentcojo, from specify
Running migrations:
  Rendering model states... DONE
  Unapplying specify.0010_updateDelete_parentcojo... OK
  Unapplying specify.0009_tectonic_ranks... OK
  Unapplying specify.0008_ageCitations_fix... OK
  Unapplying specify.0007_schema_config_update... OK
  Unapplying specify.0006_fix_tectonic_tree_fields... OK
specify@494f280856e6:/opt/specify7$ 
```

> ### Testing instructions

FYI, the "COGviewDef.txt" file for testing is not usable–
image

But I created my own 😇  which will eventually get folded into our static files in a PR I'll open for https://github.com/specify/specify7/issues/5311

If anyone else wants it, here you go: [geo.common.views.xml.zip](https://github.com/user-attachments/files/17590977/better.common.views.xml.zip)

image

> - go to data entry and select Collection Object Group 
> - define a name and select a cogType 
> - [X] verify you can create a new CO child and save the CO 
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the CO child is still there 
> - [X] verify you can create a new COG child and save the COG 
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the COG child is still there 
> - [X] verify you can add one CO child using query builder
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the CO child is still there 
> - [X] verify you can add several CO children using query builder
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the CO child is still there
> - [X] verify you can add one COG child using query builder
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the COG child is still there 
> - [X] verify you can add several COG children using query builder
> - [X] verify you can save the main COG form 
> - [X] verify after refresh the COG child is still there
> - [X] verify you can delete any children 
> - [X] verify you can uncollapse each child 
> - [ ] verify you can add a parent cojo 
> 

image

After creating a new `CollectionObjectGroupJoin` record via the `parentCojo` query combo box, I get this error upon saving:
```
BusinessRuleException at /api/specify/collectionobjectgroupjoin/ Both childcog and childco cannot be null.
```
Seems like it might not get set appropriately.

[Specify 7 Crash Report - 2024-10-31T17_14_03.769Z.txt](https://github.com/user-attachments/files/17591010/Specify.7.Crash.Report.-.2024-10-31T17_14_03.769Z.txt)


> - you should not be able to add a COGChild to a COG that is already in use in another COG, when doing so an error should be thrown saying that the COGChild is already used 

I feel pretty strongly that this error should be caught **before** crashing. No user should ever see an _error_ if we can catch it and prevent saving with a relevant tooltip.

**Review by grantfitzsimmons — APPROVED** [review]
I spoke to @CarolineDenis – the plan is to add the ability to set a `parentCojo` in January (https://github.com/specify/specify7/issues/5357)

Also I no longer get duplicate items when selecting and adding more than 5 items using the QB

**Review by acwhite211 — APPROVED** [review]

**Review by emenslin — APPROVED** [review]
### Testing instructions

- go to data entry and select Collection Object Group 
- define a name and select a cogType 
- [x] verify you can create a new CO child and save the CO 
- [x] verify you can save the main COG form 
- [x] verify after refresh the CO child is still there 
- [x] verify you can create a new COG child and save the COG 
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there 
- [x] verify you can add one CO child using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the CO child is still there 
- [x] verify you can add several CO children using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the CO child is still there
- [x] verify you can add one COG child using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there 
- [x] verify you can add several COG children using query builder
- [x] verify you can save the main COG form 
- [x] verify after refresh the COG child is still there
- [x] verify you can delete any children 
- [x] verify you can uncollapse each child 
- you should not be able to add a COGChild to a COG that is already in use in another COG, when doing so an error should be thrown saying that the COGChild is already used 

Looks good, can now add multiple CO at once!

**Comment by CarolineDenis** [comment]
Closed in favor of https://github.com/specify/specify7/pull/5360

**Pull request closed without merging**